### ISSUE TITLE
utc -> local timezone conversion fails during a DST transition

### ISSUE THREAD

**Issue by zed** [dateutil/dateutil#112]
[The bug was reported on launchpad in 2012.](https://bugs.launchpad.net/dateutil/+bug/944123) Let me know if it has been migrated to github already. I haven't found it. Here's a [standalone runnable test][1].

Input: `2011-11-06 05:30:00 UTC+0000`, `America/Toronto`
Expected: `2011-11-06 01:30:00 EDT-0400` 
Got: `2011-11-06 01:30:00 EST-0500`

The issue is reproducible with the latest `dateutil` version from PyPI.

Is it documented that the conversion should fail?
Does `dateutil` usage look ok in [the test][1]?

[1]: https://gist.github.com/zed/3838828

**Comment by pganssle** [comment]
Hm. Worth tracking down. I think, by the way, that `tz.gettz()` is the proper way to get timezone files. I'm not sure why this functionality is duplicated in `zoneinfo`.

**Comment by pganssle** [comment]
@zed FYI, the difference between the `zoneinfo` and `tz` `gettz()` functions is that the `zoneinfo` one is specifically used for getting the timezone from `dateutil`'s zoneinfo tarball.

**Comment by pganssle** [comment]
I've also looked into this actual problem a bit, and I think the problem is that we do not override the `fromutc()` function to take transition times into account. One problem with just adding `fromutc()` that the transition times are [stored as wall times](https://github.com/dateutil/dateutil/blob/master/dateutil/tz/tz.py#L397) in the `tzfile` object, so we may need to figure out how best to handle the unit conversions.

**Comment by pganssle** [comment]
On second thought, according to [this comment](https://github.com/python/cpython/blob/8d1d7e6816753248768e4cc1c0370221814e9cf1/Lib/datetime.py#L1954) (which, frankly, I'm having a bit of trouble following, since I'm not really sure why the notation makes a distinction between `y.n` and `x.n`), overriding `fromutc()` is generally unnecessary, and looking at this more closely, I see what the problem is.

This relates to the fact that `pytz` eagerly evaluates the DST status of the timezone, which allows them to handle ambiguous date representations for the user. In a sense they are extending `datetime` using `tzinfo` objects. This allows them to correctly handle DST status of "nominally" ambiguous times (e.g. times where, in the specified zone, the date represented by `dt.replace(tzinfo=None)` is ambiguous), but causes problems with datetime arithmetic, since they don't re-evaluate the DST status in response to changes in the datetime.

`dateutil` on the other hand, lazily evaluates the time zone as necessary, and as such our timezones handle arithmetic fine, but we can't handle "nominally" ambiguous times, because each time we calculate the DST status, we're doing it 100% based on the (ambiguous) naive datetime object.

This will be solved when [PEP495](https://www.python.org/dev/peps/pep-0495/) is implemented and released, because at that point, `datetime` objects will no longer be underspecified with respect to daylight savings.

I see two possible paths forward:

1. Do nothing, because time zones probably shouldn't be stateful in the first place, and wait for PEP495 to solve things for us.
2. Implement as a stop-gap measure a stateful timezone object that (as lightly as possible) overrides `fromutc()`, to allow us to store a "fold" flag in our `tzinfo` classes. This would not work perfectly with arithmetic, but I'm thinking that the "best approximation" would look like this:
    * When `fromutc()` is called, check what side of the DST line we're on and set a `_fold` property appropriately in the `tzinfo` object.
    * When `utcoffset()` (or `tzname()` or anything else requiring that we resolve the actual time zone) is called, check to see if we're at an ambiguous time, then:
        * If we are, resolve in favor of the direction indicated by the `_fold` property.
        * If we're not, set the `_fold` property to whichever side of the fold we're on.
  That will still give some incorrect results when you're applying arithmetic operations and landing in ambiguous times, but I think that that's just something that can't be helped, and this will work fine for most cases.

I'm thinking that there's nothing horribly pathological about the second one. Maybe @abalkin would be willing to comment?

**Comment by abalkin** [comment]
I definitely recommend the first option.  The second option looks somewhat like the pytz kludge which was much frowned upon during PEP 495 discussions over at the datetime-sig mailing list.

**Comment by pganssle** [comment]
@abalkin My instinct is to agree (which is why it's the first option), and I'll review the datetime-sig mailing list options, but I suspect that it will be many, many years before we can drop support for pre-PEP 495 versions of Python.

I think in the interests of practicality, a "light touch" kludge might be justified. The `pytz` kludge is quite heavy-handed, whereas I think my proposed solution is a reasonable heuristic given how `tzinfo` objects are actually used.

**Comment by abalkin** [comment]
@pganssle - Can you describe your solution in more detail?  Note that typically tzinfo objects are shared between datetime instances.  Even if you don't implement them as singletons, whenever you add a timedelta to a datetime instance, you get two datetimes with the same tzinfo.  Now, if you call fromutc() on one of those instances would both get the _fold property?

The only workable solution that I can think of would be to maintain a list (maybe in a weakref.WeakKeyDictionary) of datetime instances that in PEP 495 world would have fold=1.  The fromutc()  can then add to this list and utcoffset consult this list in the ambiguous cases.

**Comment by pganssle** [comment]
@abalkin Well, trying to work out the details, but I think that `fromutc()` is the only time you can absolutely know whether you're in a fold, and the idea would be to bias future ambiguous dates in favor of having the same fold-resolution behavior. Your point is well taken about `tzinfo` objects being generally shared, so I think the `utcoffset()` calls should not alter the state. Here's a bit of rough code outlining how I might implement it:

```python
class _tzinfo(datetime.tzinfo):
    """
    Base class for all `dateutil` `tzinfo` objects.
    """
    
    def __init__(*args, **kwargs):
        super(_tzinfo, self).__init__(*args, **kwargs)

        self._fold = None

    def fromutc(self, dt):
        """
        Convert the specified datetime into 
        """

        _fold_cached = self._fold
        self._fold = None

        ndt = super(_tzinfo, self).fromutc(dt)

        _fold = (ndt - dt != (dt.utcoffset() - dt.dst()))

        self._fold = _fold_cached

        if self._fold is None:
            self._fold = _fold
        elif _fold != self._fold:
            newtz = deepcopy(self)
            newtz._fold = _fold
            ndt = ndt.replace(tzinfo=newtz)

        return ndt
```

Then the individual subclasses would do with the `_fold` information whatever they're going to do with it. Presumably I'd actually wrap the retrieval of the `_fold` attribute in a method, so that when PEP 495 rolls out, I can do something like this:

```python
    def _get_fold(self, dt):
        try:
            return dt.fold
        except AttributeError:
            return self._fold
```

**Comment by pganssle** [comment]
@abalkin My first pass at an implementation for this is PR #225, if you have any input.

**Comment by abalkin** [comment]
@pganssle - please take a look at ZoneInfoCompleteTest in python/cpython#20.  I think it is not hard to adapt it for dateutil.  If your work-around passes this test - I'll be quite impressed.

**Comment by pganssle** [comment]
@abalkin Nice. I took a quick crack at adapting it, the `folds()` test is passing for all zones, but the `gaps()` isn't (which is sorta what I was expecting, but something to look into). I'll play around with it a bit later.

**Comment by pganssle** [comment]
Fixed in #225.
